packages/wallet/src/validator/rollup-fraud-guard.ts

```diff
@@ -0,0 +1,162 @@
+ */
+export class RollupFraudGuard implements EthereumListener<Event> {
+  public static readonly LAST_BLOCK_VALIDATED_KEY = Buffer.from(
+    'LAST_VALIDATED_BLOCK'
+  )
+  private static readonly lockKey: string = 'lock'
+
+  private readonly lock: AsyncLock
+  private lastBlockValidated: number
+
+  public static async create(
+    db: DB,
+    validator: RollupStateValidator,
+    contract: Contract
+  ): Promise<RollupFraudGuard> {
+    const fraudGuard: RollupFraudGuard = new RollupFraudGuard(
+      db,
+      validator,
+      contract
+    )
+
+    await fraudGuard.init()
+
+    return fraudGuard
+  }
+
+  private constructor(
+    private readonly db: DB,
+    private readonly validator: RollupStateValidator,
+    private readonly contract: Contract
+  ) {
+    this.lock = new AsyncLock()
+  }
+
+  private async init(): Promise<void> {
+    const lastValidatedBuffer: Buffer = await this.db.get(
+      RollupFraudGuard.LAST_BLOCK_VALIDATED_KEY
+    )
+
+    if (!lastValidatedBuffer) {
+      log.info(`Starting fresh. Last validated block: 0`)
+      this.lastBlockValidated = 0
+      return
+    }
+
+    this.lastBlockValidated = parseInt(lastValidatedBuffer.toString(), 10)
+    log.info(
+      `Starting from previous run. Last validated block: ${this.lastBlockValidated}`
+    )
+  }
+
+  public async onSyncCompleted(syncIdentifier?: string): Promise<void> {
+    // no-op
+  }
+
+  public async handle(event: Event): Promise<void> {
+    log.debug(`Fraud Guard received event: ${JSON.stringify(event)}`)
+    if (
+      !event ||
+      (!event.values &&
+        !('block' in event.values) &&
+        !('blockNumber' in event.values))
+    ) {
+      log.error(`Unrecognized event. Returning`)
+      return
+    }
+
+    let block: RollupBlock
+    try {
+      block = {
+        blockNumber: (event.values['blockNumber'] as any).toNumber(),
+        transitions: (event.values['block'] as string[]).map((x) =>
+          parseTransitionFromABI(x)
+        ),
+      }
+
+      return this.lock.acquire(RollupFraudGuard.lockKey, async () => {
+        await this.handleNewRollupBlock(block)
+      })
+    } catch (e) {
+      logError(
+        log,
+        `Error trying to parsing and handling event: ${JSON.stringify(event)}`,
+        e
+      )
+      return
+    }
+  }
+
+  private async handleNewRollupBlock(block: RollupBlock): Promise<void> {
+    if (this.lastBlockValidated >= block.blockNumber) {
+      log.debug(
+        `Received event for old block. Ignoring. lastValidated: ${
+          this.lastBlockValidated
+        }. Received: ${JSON.stringify(block)}`
+      )
+      return
+    } else if (this.lastBlockValidated + 1 !== block.blockNumber) {
+      log.error(
+        `Received event with block number greater than expected! lastValidated: ${
+          this.lastBlockValidated
+        }. Received: ${JSON.stringify(block)}`
+      )
+      process.exit(1)
```

Comment:
Is there any chance that Ethereum events will come in out of order?  I'm not super familiar with ethereum event logging, so may be a silly question, but thought I'd mention because this would cause issues if that were to happen.  (and if it can happen but is unlikely, seems low priority to solve)